### PR TITLE
two minor memleak fixes for sysusers, fixing #8718

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/sysusers/sysusers.c [hunk 1]
```diff
@@ -112,12 +112,12 @@ static int load_user_database(void) {
...
                 q = hashmap_put(database_uid, UID_TO_PTR(pw->pw_uid), n);
                 if (q < 0 && q != -EEXIST) {
-                        if (k < 0)
+                        if (k <= 0)
                                 free(n);
                         return q;
                 }
 
-                if (q < 0 && k < 0)
+                if (q <= 0 && k <= 0)
```
Comment (evverx) [inline src/sysusers/sysusers.c:120]: I'm not sure if it's possible for `q` to be equal to `0` here, but if, for whatever reason, this happens, it seems that the memory `n` points to will be freed at least twice. To prevent it, I think `hashmap_remove` should be called before `free`.

Comment (evverx) [reply src/sysusers/sysusers.c:120]: `hashmap_remove` might help avoid calling `free`, but it also would make `systemd-sysusers` incorrect, so it seems that something else should be done.

File: src/sysusers/sysusers.c [hunk 2]
```diff
@@ -97,8 +97,8 @@ static int load_user_database(void) {
                 return r;
 
         while ((r = fgetpwent_sane(f, &pw)) > 0) {
-                char *n;
                 int k, q;
+                char *n;
```
Comment (keszybz) [inline src/sysusers/sysusers.c:101]: Considering that `n` is assigned first, I don't see why this change.

File: src/sysusers/sysusers.c (hunk 1 above)
Comment (keszybz) [reply src/sysusers/sysusers.c:120]: If `q <= 0 && k <= 0`, `n` wasn't added anywhere, and can be freed safely. This updated conditions appears correct.

Comment (evverx) [reply src/sysusers/sysusers.c:120]: I think `q < 0 && k <= 0` is correct. `q == 0` seems to be something unexpected meaning that the address of the memory `n` points to has already been stored in `database_uid`. That memory is freed here and then in `free_database`, and also read somewhere in between.

Comment (evverx) [reply src/sysusers/sysusers.c:120]: > If q <= 0 && k <= 0, n wasn't added anywhere, and can be freed safely

I'm not sure I get it. Could you expand on this? This particular check is supposed to make sure that the memory is freed if it's the second time a user or an id has been encountered.

**Comment by evverx** [comment]
I think it would also be great if `systemd-sysusers` could remove the duplicates it created due to  https://github.com/systemd/systemd/issues/8721. Should I open another issue or is it expected that `systemd-sysusers` keeps duplicates?

File: src/sysusers/sysusers.c (hunk 1 above)
Comment (keszybz) [reply src/sysusers/sysusers.c:120]: `q == 0` is not possible, because that'd mean that the address that `strdup` returned was already used before. So `q < 0 && k <= 0` would be correct, but `q <= 0 && k <= 0` is completely equivalent, *and* I think it's easier to understand.

I'll push a trivial followup to order `q` and `k` consistently though.

Comment (evverx) [reply src/sysusers/sysusers.c:120]: `q == 0 && k == 0` doesn't seem to be possible now just because there aren't any errors. It would be possible, though, if there were a bug somewhere. I believe `assert_se(q != 0 && k != 0)` would make it more clear that the condition isn't considered to be possible.

Comment (keszybz) [reply src/sysusers/sysusers.c:120]: `k == 0` is possible.

Comment (evverx) [reply src/sysusers/sysusers.c:120]: That's true. Anyway, my point is that if something isn't considered to be possible and clearly indicates a programming error, it should probably be wrapped in `assert` or whatever else is used to detect programming errors.

**Comment by poettering** [comment]
> I think it would also be great if systemd-sysusers could remove the duplicates it created due to #8721. Should I open another issue or is it expected that systemd-sysusers keeps duplicates?

Hmm, I wonder if it's worth it, this was only in for a very short time... It's ugly, but if something cleans this up it should probably be in postinst RPM scripts or so, not in C code...
